tests/unit_tests/homeassistant/test_online_sensor.py

```diff
@@ -105,4 +113,7 @@ async def test_get_value_off(monkeypatch):
         "id": 12345678901234567,
         "vehicle_id": 1234567890,
         "update_interval": 300,
-    }
+        "vehicle_data": '{"climate_state": {}, "charge_state": {}, "vehicle_state": '
```

Comment:
Please have the test mock some test data in addition to the empty data set.

Reply:
Updated test but seem to have issues with mocking the one with data.

Reply:
The tests indicate there is already mock data.

```
E         Full diff:
E           {
E            'id': 12345678901234567,
E            'state': 'asleep',
E            'update_interval': 300,
E         -  'vehicle_data': '{"climate_state": {"climate_state": "on"}, "charge_state": '
E         +  'vehicle_data': '{"climate_state": {"5YJSA11111111111": {"climate_state": '
E         +                  '"on"}}, "charge_state": {"5YJSA11111111111": '
E         -                  '{"charge_state": "on"}, "vehicle_state": {"vehicle_state": '
E         ?                          ^                                 ------------------
E         +                  '{"charging_state": "on"}}, "vehicle_state": '
E         ?                          ^^^              +
E         +                  '{"5YJSA11111111111": {"vehicle_state": "on"}}, '
E         -                  '"on"}, "vehicle_config": {"vehicle_config": "on"}, '
E         ?                   -------                                     -------
E         +                  '"vehicle_config": {"5YJSA11111111111": {"vehicle_config": '
E         ?                                    +++++++++++++++++++++
E         -                  '"drive_state": {"drive_state": "on"}, "gui_settings": '
E         -                  '{"gui_settings": "on"}',
E         +                  '"on"}}, "drive_state": {"5YJSA11111111111": {"drive_state": '
E         +                  '"on"}}, "gui_settings": {"5YJSA11111111111": {"gui_setting": '
E         +                  '"on"}}}',
```
It's likely from this line. https://github.com/zabuldon/teslajsonpy/pull/279/files#diff-033c34e16f4931eef5f977af4319399de9fd75382786738a2c80a2648ea6c619R96

You should be modifying _data like these tests: https://github.com/zabuldon/teslajsonpy/blob/e2dfa6a41f2a1be260bf89065751c1359564792e/tests/unit_tests/homeassistant/test_online_sensor.py#L68

You should have tests that cover:

1. Test for empty data attribute (for example if that values aren't set because an update fails.
2. Test for some actual data.

Ideally these tests should be self contained compared to other tests so if one fails, we can zero in on what behavior is broken. If it's easy to add them as part of an existing test, then that's fine as long as the logic is clear.

Reply:
Empty is tested on both test_get_value_on and test_get_value_off. Added a new test test_get_vehicle_data_attribute to specifically test when attributes are filled in.